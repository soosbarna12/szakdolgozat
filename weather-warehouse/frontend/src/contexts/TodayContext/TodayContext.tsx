import React, { createContext, useEffect, useState } from "react";
import { TodayContextType, TodayLocationData, TodayProviderProps } from "./TodayContext.type";


export const TodayContext = createContext<TodayContextType>({
	location: { name: "", lat: 0, lon: 0 },
	setLocation: () => { },
});

export const TodayLocationProvider = ({ children }: TodayProviderProps) => {

	const [location, setLocation] = useState<TodayLocationData>(() => {
		const storedLocation = localStorage.getItem("todayLocation");
		const defaultLocation = { name: "", lat: 0, lon: 0 };
		try {
			return storedLocation ? JSON.parse(storedLocation) : defaultLocation;
		} catch (error) {
			return defaultLocation;
		}
	});

	useEffect(() => {
		localStorage.setItem("todayLocation", JSON.stringify(location));
	}, [location]);


	return (
		<TodayContext.Provider value={{ location, setLocation }}>
			{children}
		</TodayContext.Provider>
	);
};
