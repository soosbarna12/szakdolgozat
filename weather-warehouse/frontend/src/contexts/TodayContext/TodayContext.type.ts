import { Dispatch, ReactNode, SetStateAction } from "react";

// location name interfere
export interface TodayLocationData {
  name: string;
  country?: string;
  state?: string;
  lat: number;
  lon: number;
}

export interface TodayContextType {
  location: TodayLocationData;
  setLocation: Dispatch<SetStateAction<TodayLocationData>>;
}

export interface TodayProviderProps {
	children: ReactNode;
}
