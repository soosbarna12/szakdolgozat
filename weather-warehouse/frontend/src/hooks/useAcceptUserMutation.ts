import { useMutation } from "@tanstack/react-query";
import { useAlert } from "../utils/AlertContext";
import axios from "../utils/axiosConfig";


export function useAcceptUserMutation() {
  const { showAlert } = useAlert();

  const mutation = useMutation({
    mutationFn: async (id: number) => {
      const response = await axios.post(`/user/accept`, { id });
      return response.data;
    },
    onSuccess: () => {
      showAlert("User accepted successfully", "success");
    },
    onError: () => {
      showAlert("Failed to accept user", "error");
    },
  });

  return mutation;
}
