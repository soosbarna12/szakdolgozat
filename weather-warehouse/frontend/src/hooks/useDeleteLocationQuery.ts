import { useMutation } from "@tanstack/react-query";
import { useAlert } from "../utils/AlertContext";
import axios from "../utils/axiosConfig";

export function useDeleteLocationQuery() {
  const { showAlert } = useAlert();

  const mutation = useMutation({
    mutationFn: async (userLocationID: number) => {
      const response = await axios.delete(`/user/deleteLocation`, { params: { userLocationID } });
      return response.data;
    },
    onSuccess: () => {
      showAlert("Location deleted successfully", "success");
    },
    onError: () => {
      showAlert("Failed to delete location", "error");
    },
  });

  return mutation;
}