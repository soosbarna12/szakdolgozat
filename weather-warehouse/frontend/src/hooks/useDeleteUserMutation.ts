import { useMutation } from "@tanstack/react-query";
import { useAlert } from "../utils/AlertContext";
import axios from "../utils/axiosConfig";


export function useDeleteUserMutation() {
  const { showAlert } = useAlert();

  const mutation = useMutation({
    mutationFn: async (id: number) => {
      const response = await axios.delete(`/user/delete`, { params: { id } });
      return response.data;
    },
    onSuccess: () => {
      showAlert('User deleted successfully', 'success');
    },
    onError: () => {
      showAlert('Failed to delete user', 'error');
    },
  });

  return mutation;
}
