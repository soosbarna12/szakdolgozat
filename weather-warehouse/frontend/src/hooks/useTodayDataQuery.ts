import { useQuery } from "@tanstack/react-query";
import { useAlert } from "../utils/AlertContext";
import axios from "axios";


export function useTodayDataQuery(lat: number, lon: number) {
  const { showAlert } = useAlert();
  
  const { data, error, isLoading } = useQuery({
    queryKey: ["todayWeatherData", lat, lon],
    queryFn: async () => {
      const response = await axios.get(`/today/locationData`, {
        //params: { location, lang: "en" },
        params: { lat, lon, lang: "en" },
      });

      if (response.data?.cod === "404") {
        throw new Error("City not found");
      }

      return response.data;
    },
    enabled: !!(lat && lon)
  });

  if (error) {
    showAlert("Error fetching today's weather data", "error");
  }

  return { data, error, isLoading };
}
