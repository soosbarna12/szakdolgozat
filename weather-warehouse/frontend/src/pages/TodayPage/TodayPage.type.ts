export interface ProfileMenuProps {
	handleSetTheme: (newTheme: boolean) => void;
	handleLocationChange: (newLocation: string) => void;
}
