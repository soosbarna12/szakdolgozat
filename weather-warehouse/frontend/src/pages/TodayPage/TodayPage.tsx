import { Skeleton } from "@mui/material";
import React, { useContext } from "react";
import { TodayWeatherCard } from "../../components/DataGrids/WeatherCard/TodayWeatherCard";
import { FilterBar } from "../../components/FilterBar/FilterBar";
import { TodayContext } from "../../contexts/TodayContext/TodayContext";
import { useTodayDataQuery } from "../../hooks/useTodayDataQuery";
import { ContentBox, StyledItem } from "../../stlyes/content.style";
import { Pages } from "../../types/page.type";


export function TodayPage() {
  const { location } = useContext(TodayContext);
  //console.log("TodayPage location: ", location);
  //const { data: todayData, error, isLoading } = useTodayDataQuery(location.lat, location.lon); // currently using the todays data query, because the historical is not available yet
  const { data: todayData, error, isLoading } = useTodayDataQuery(location.name); // currently using the todays data query, because the historical is not available yet


  function renderShimmer() {
    if (isLoading) {
      return (
        <Skeleton
          variant="rectangular"
          animation="wave"
          width="350px"
          height="375px"
          data-testid="todaySkeleton"
          sx={{ borderRadius: '20px' }}
        />
      )
    }
  }

  function renderError() {
    if (error) {
      return <p>Error fetching data.</p>;
    }
  }

  function renderWeatherCard() {
    if (todayData) {
      return (
        <StyledItem sx={{ height: "400px", width: "350px" }} data-testid="todayWeatherCard">
          <TodayWeatherCard data={todayData} />
        </StyledItem>
      );
    }
  }

  return (
    <>
      <FilterBar type={Pages.Today} location={location.name} />
      <ContentBox sx={{ display: "flex", justifyContent: "center" }}>
        {renderShimmer()}
        {renderError()}
        {renderWeatherCard()}
      </ContentBox>
    </>
  );
}
