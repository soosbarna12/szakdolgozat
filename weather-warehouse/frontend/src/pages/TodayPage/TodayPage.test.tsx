import '@testing-library/jest-dom';
import { screen } from '@testing-library/react';
import React from 'react';
import { TodayContext } from '../../contexts/TodayContext/TodayContext';
import { renderWithQueryClient } from '../../utils/test/renderWithQueryClient';
import { TodayPage } from './TodayPage';
import { useTodayDataQuery } from '../../hooks/useTodayDataQuery';

jest.mock('../../hooks/useTodayDataQuery', () => ({
  useTodayDataQuery: jest.fn(),
}));

jest.mock('react-leaflet', () => ({
  MapContainer: ({ children }: { children: React.ReactNode }) => (
    <div data-testid="dataMapContainer">{children}</div>
  ),
  TileLayer: () => <div data-testid="mockTileLayer">Mock TileLayer</div>,
  useMap: jest.fn(() => ({
    setView: jest.fn(),
  })),
}));

describe('pages/TodayPage', () => {
  const mockLocation = { name: 'New York', lat: 40.7128, lon: -74.006 };
  const mockTodayData = {
    name: 'New York',
    sys: { country: 'US' },
    dt: 1633036800,
    weather: [{ main: 'Clear', description: 'clear sky' }],
    main: { temp: 25, humidity: 60 },
    wind: { speed: 5 },
  };
  const mockError = 'Error fetching data';

  const renderTodayPage = () =>
    renderWithQueryClient(
      <TodayContext.Provider value={{ location: mockLocation } as any}>
        <TodayPage />
      </TodayContext.Provider>
    );

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('matches TodayPage component snapshot', () => {
    (useTodayDataQuery as jest.Mock).mockReturnValue({
      data: mockTodayData,
      error: null,
      isLoading: false,
    });

    const { container } = renderTodayPage();
    expect(container).toMatchSnapshot();
  });

  it('renders shimmer when loading', () => {
    (useTodayDataQuery as jest.Mock).mockReturnValue({
      data: null,
      error: null,
      isLoading: true,
    });

    renderTodayPage();

    const shimmer = screen.getByTestId('todaySkeleton');
    expect(shimmer).toBeInTheDocument();
  });

  it('renders error message when there is an error', () => {
    (useTodayDataQuery as jest.Mock).mockReturnValue({
      data: null,
      error: mockError,
      isLoading: false,
    });

    renderTodayPage();

    const errorMessage = screen.getByText('Error fetching data.');
    expect(errorMessage).toBeInTheDocument();
  });

  it('renders weather card when data is available', () => {
    (useTodayDataQuery as jest.Mock).mockReturnValue({
      data: mockTodayData,
      error: null,
      isLoading: false,
    });

    renderTodayPage();

    const weatherCard = screen.getByTestId('todayWeatherCard');
    expect(weatherCard).toBeInTheDocument();
  });

  it('handles null data gracefully', () => {
    (useTodayDataQuery as jest.Mock).mockReturnValue({
      data: null,
      error: null,
      isLoading: false,
    });

    renderTodayPage();

    const weatherCard = screen.queryByTestId('todayWeatherCard');
    expect(weatherCard).toBeNull();
  });
});