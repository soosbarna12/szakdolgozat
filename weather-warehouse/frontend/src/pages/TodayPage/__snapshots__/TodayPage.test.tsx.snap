// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/TodayPage matches TodayPage component snapshot 1`] = `
<div>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation0 MuiAppBar-root MuiAppBar-colorTransparent MuiAppBar-positionStatic css-94iixj-MuiPaper-root-MuiAppBar-root"
    style="--Paper-shadow: none;"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-t4zkma-MuiToolbar-root"
    >
      <div
        class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-1tlcqt-MuiAutocomplete-root"
      >
        <div
          class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-rs1at4-MuiFormControl-root-MuiTextField-root"
          data-testid="locationSearchBox"
        >
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-1e3qgpi-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1nowbqt-MuiInputAdornment-root"
            >
              <span
                aria-hidden="true"
                class="notranslate"
              >
                ​
              </span>
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="SearchOutlinedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
                />
              </svg>
            </div>
            <input
              aria-autocomplete="list"
              aria-expanded="false"
              aria-invalid="false"
              autocapitalize="none"
              autocomplete="off"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-qve9l1-MuiInputBase-input-MuiOutlinedInput-input"
              id="«r0»"
              placeholder="New York"
              role="combobox"
              spellcheck="false"
              type="text"
              value=""
            />
            <div
              class="MuiAutocomplete-endAdornment css-1uhhrmm-MuiAutocomplete-endAdornment"
            >
              <button
                aria-label="Open"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-2y136s-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                tabindex="-1"
                title="Open"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
              </button>
            </div>
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1ll44ll-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-w4cd9x"
              >
                <span
                  aria-hidden="true"
                  class="notranslate"
                >
                  ​
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
    </div>
  </header>
  <div
    class="MuiBox-root css-1v3ilv5"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-osau60-MuiPaper-root"
      data-testid="todayWeatherCard"
      style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
    >
      <div
        class="MuiBox-root css-13brihr"
      >
        <h6
          class="MuiTypography-root MuiTypography-h6 css-1miy0lu-MuiTypography-root"
          data-testid="date"
        >
          2021. szeptember 30., csütörtök
        </h6>
        <h5
          class="MuiTypography-root MuiTypography-h5 css-stnzmu-MuiTypography-root"
          data-testid="location"
        >
          New York
          , 
          US
        </h5>
        <div
          class="MuiBox-root css-nkffzc"
        >
          <div
            class="MuiBox-root css-1niqckn"
            data-testid="weatherIcon"
          >
            <svg
              data-testid="wiDaySunny"
              fill="currentColor"
              height="120px"
              id="Layer_1"
              stroke="currentColor"
              stroke-width="0"
              version="1.1"
              viewBox="0 0 30 30"
              width="120px"
              x="0px"
              xmlns="http://www.w3.org/2000/svg"
              y="0px"
            >
              <path
                d="M4.37,14.62c0-0.24,0.08-0.45,0.25-0.62c0.17-0.16,0.38-0.24,0.6-0.24h2.04c0.23,0,0.42,0.08,0.58,0.25
	c0.15,0.17,0.23,0.37,0.23,0.61S8,15.06,7.85,15.23c-0.15,0.17-0.35,0.25-0.58,0.25H5.23c-0.23,0-0.43-0.08-0.6-0.25
	C4.46,15.06,4.37,14.86,4.37,14.62z M7.23,21.55c0-0.23,0.08-0.43,0.23-0.61l1.47-1.43c0.15-0.16,0.35-0.23,0.59-0.23
	c0.24,0,0.44,0.08,0.6,0.23s0.24,0.34,0.24,0.57c0,0.24-0.08,0.46-0.24,0.64L8.7,22.14c-0.41,0.32-0.82,0.32-1.23,0
	C7.31,21.98,7.23,21.78,7.23,21.55z M7.23,7.71c0-0.23,0.08-0.43,0.23-0.61C7.66,6.93,7.87,6.85,8.1,6.85
	c0.22,0,0.42,0.08,0.59,0.24l1.43,1.47c0.16,0.15,0.24,0.35,0.24,0.59c0,0.24-0.08,0.44-0.24,0.6s-0.36,0.24-0.6,0.24
	c-0.24,0-0.44-0.08-0.59-0.24L7.47,8.32C7.31,8.16,7.23,7.95,7.23,7.71z M9.78,14.62c0-0.93,0.23-1.8,0.7-2.6s1.1-1.44,1.91-1.91
	s1.67-0.7,2.6-0.7c0.7,0,1.37,0.14,2.02,0.42c0.64,0.28,1.2,0.65,1.66,1.12c0.47,0.47,0.84,1.02,1.11,1.66
	c0.27,0.64,0.41,1.32,0.41,2.02c0,0.94-0.23,1.81-0.7,2.61c-0.47,0.8-1.1,1.43-1.9,1.9c-0.8,0.47-1.67,0.7-2.61,0.7
	s-1.81-0.23-2.61-0.7c-0.8-0.47-1.43-1.1-1.9-1.9C10.02,16.43,9.78,15.56,9.78,14.62z M11.48,14.62c0,0.98,0.34,1.81,1.03,2.5
	c0.68,0.69,1.51,1.04,2.49,1.04s1.81-0.35,2.5-1.04s1.04-1.52,1.04-2.5c0-0.96-0.35-1.78-1.04-2.47c-0.69-0.68-1.52-1.02-2.5-1.02
	c-0.97,0-1.8,0.34-2.48,1.02C11.82,12.84,11.48,13.66,11.48,14.62z M14.14,22.4c0-0.24,0.08-0.44,0.25-0.6s0.37-0.24,0.6-0.24
	c0.24,0,0.45,0.08,0.61,0.24s0.24,0.36,0.24,0.6v1.99c0,0.24-0.08,0.45-0.25,0.62c-0.17,0.17-0.37,0.25-0.6,0.25
	s-0.44-0.08-0.6-0.25c-0.17-0.17-0.25-0.38-0.25-0.62V22.4z M14.14,6.9V4.86c0-0.23,0.08-0.43,0.25-0.6C14.56,4.09,14.76,4,15,4
	s0.43,0.08,0.6,0.25c0.17,0.17,0.25,0.37,0.25,0.6V6.9c0,0.23-0.08,0.42-0.25,0.58S15.23,7.71,15,7.71s-0.44-0.08-0.6-0.23
	S14.14,7.13,14.14,6.9z M19.66,20.08c0-0.23,0.08-0.42,0.23-0.56c0.15-0.16,0.34-0.23,0.56-0.23c0.24,0,0.44,0.08,0.6,0.23
	l1.46,1.43c0.16,0.17,0.24,0.38,0.24,0.61c0,0.23-0.08,0.43-0.24,0.59c-0.4,0.31-0.8,0.31-1.2,0l-1.42-1.42
	C19.74,20.55,19.66,20.34,19.66,20.08z M19.66,9.16c0-0.25,0.08-0.45,0.23-0.59l1.42-1.47c0.17-0.16,0.37-0.24,0.59-0.24
	c0.24,0,0.44,0.08,0.6,0.25c0.17,0.17,0.25,0.37,0.25,0.6c0,0.25-0.08,0.46-0.24,0.62l-1.46,1.43c-0.18,0.16-0.38,0.24-0.6,0.24
	c-0.23,0-0.41-0.08-0.56-0.24S19.66,9.4,19.66,9.16z M21.92,14.62c0-0.24,0.08-0.44,0.24-0.62c0.16-0.16,0.35-0.24,0.57-0.24h2.02
	c0.23,0,0.43,0.09,0.6,0.26c0.17,0.17,0.26,0.37,0.26,0.6s-0.09,0.43-0.26,0.6c-0.17,0.17-0.37,0.25-0.6,0.25h-2.02
	c-0.23,0-0.43-0.08-0.58-0.25S21.92,14.86,21.92,14.62z"
              />
            </svg>
          </div>
          <h2
            class="MuiTypography-root MuiTypography-h2 css-6fhuif-MuiTypography-root"
            data-testid="temperature"
          >
            -248
            <span
              style="font-size: 0.5em; vertical-align: super;"
            >
               
              °C
            </span>
          </h2>
        </div>
        <h6
          class="MuiTypography-root MuiTypography-h6 css-pd6es5-MuiTypography-root"
          data-testid="desciption"
        >
          Clear sky
        </h6>
        <div
          class="MuiBox-root css-0"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
            data-testid="feelsLike"
          >
            Feels Like: 
            NaN
             
            °C
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
            data-testid="precipitation"
          >
            Precipitation: 
            0
             mm/h
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
            data-testid="humidity"
          >
            Humidity: 
            60
            %
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
            data-testid="pressure"
          >
            Pressure: 
             hPa
          </p>
          <p
            class="MuiTypography-root MuiTypography-body1 css-rizt0-MuiTypography-root"
            data-testid="windSpeed"
          >
            Wind: 
            5
             m/s
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
